tmt/steps/prepare/distgit.py

```diff
@@ -0,0 +1,293 @@
+            # Install build requires
+            # Create the package manager command
+            cmd, _ = guest.package_manager.prepare_command()
+            # Can't set 'cwd' as the check for its existence fails for local workdir
+            cmd += Command("builddep", "-y", f"SRPMS/{src_rpm_name}")
+            guest.execute(command=cmd, cwd=Path(source_dir))
+
+        # Finally run the rpm-build -bp
+        cmd = Command(
+            "rpmbuild", "-bp", spec_name, "--nodeps",
+            *dir_defines
+            )
+        try:
+            guest.execute(command=cmd,
+                          cwd=source_dir,
+                          )
+        except tmt.utils.RunError as error:
+            raise tmt.utils.PrepareError("Unable to 'rpmbuild -bp'.", causes=[error])
+
+        # Make sure to pull back sources ...
+        # FIXME -- Do we need to? Can be lot of data...
+        guest.pull(source_dir)
+
+        # Mark which file/dirs were created after rpmbuild ran
+        content_after = set(source_dir.iterdir())
+        created_content = [Path(p.name) for p in content_after - content_before]
+
+        # When discover is set let it rediscover tests
+        if self.discover is not None:
+            self.discover.post_dist_git(created_content)
+            # FIXME needs refactor of Prepare, tmt.base etc...
+            # doing quick & dirty injection of prepareinstalls
+            for guest in self.step.plan.provision.guests():
+                collected_requires = []
+                collected_recommends = []
+                for test in self.step.plan.discover.tests(enabled=True):
+                    if not test.enabled_on_guest(guest):
+                        continue
+
+                    collected_requires += tmt.base.assert_simple_dependencies(
+                        test.require,
+                        'After beakerlib processing, tests may have only simple requirements',
+                        self._logger)
+
+                    collected_recommends += tmt.base.assert_simple_dependencies(
+                        test.recommend,
+                        'After beakerlib processing, tests may have only simple requirements',
+                        self._logger)
+
+                    collected_requires += test.test_framework.get_requirements(
+                        test,
+                        self._logger)
+
+                    for check in test.check:
+                        collected_requires += check.plugin.essential_requires(
+                            guest,
+                            test,
+                            self._logger)
+                # Inject additional install plugins - require
+                if collected_requires and self.future_requires:
+                    self.future_requires.data.package = [
+                        dependency.to_spec() for dependency in collected_requires]
```

Comment:
The `data.package` is a list of `DependencySimple` so we should not convert to spec, right? I guess, we should also remove duplicate dependencies. What about something like this?

```suggestion
                    self.future_requires.data.package = list(set(collected_requires))
```

This also does not handle packages to be installed on proper guests, but I'd say as a quick & dirty fix sounds good for now, we can clean this up later. @happz, any quick suggestions?

Reply:
I copied the line from https://github.com/teemtee/tmt/blob/4ad9ce65494dd3743fbd9aec298b90ec9f25c900/tmt/steps/prepare/__init__.py#L311 as well as other lines from prepare/__init__.py. 
Does it behave differently in this plugin then?

> This also does not handle packages to be installed on proper guests,

Why it doesn't, The `if not test.enabled_on_guest(guest):` check doesn't work?

Reply:
@lukaszachy 

> Does it behave differently in this plugin then?

Yes, it does. In `__init__.py`, it was put into a "raw step data" for, which is a dictionary one might get from an fmf node, which was then passed to `delegate()` - a plugin is constructed, raw step data normalized, "spec" turns into `Dependency*`. You skip this part, construct the plugin instance on your own, and then inject it into its step data container. The raw step data you have do not get normalized.

Reply:
> Does it behave differently in this plugin then?

Here the raw step data are constructed:

https://github.com/teemtee/tmt/blob/4ad9ce65494dd3743fbd9aec298b90ec9f25c900/tmt/steps/prepare/__init__.py#L305-L315

While in the dist-git code `self.future_requires.data.package` is already the real data, the plugin data object which does not hold specification (e.g. strings and keys from `fmf` files) but real objects.

Reply:
> Why it doesn't, The `if not test.enabled_on_guest(guest):` check doesn't work?

Inserting prepare phase to relevant guests only is handled here:

https://github.com/teemtee/tmt/blob/4ad9ce65494dd3743fbd9aec298b90ec9f25c900/tmt/steps/prepare/__init__.py#L301-L315

Which means checking each test and storing the information about the guest on which the test is enabled first and only then creating appropriate prepare phases.

Reply:
> Yes, it does. In `__init__.py`, it was put into a "raw step data" for, which is a dictionary one might get from an fmf node, which was then passed to `delegate()` - a plugin is constructed, raw step data normalized, "spec" turns into `Dependency*`. You skip this part, construct the plugin instance on your own, and then inject it into its step data container. The raw step data you have do not get normalized.

"You skip this part," I'm not, to my knowledge, see
```
    future_recommends: PreparePlugin[Any] = cast(
        PreparePlugin[Any], PreparePlugin.delegate(
            prepare_step, raw_data=data_recommend))
    prepare_step._phases.append(future_recommends)
```

Reply:
@psss IIRC this PR was for single host scenario (which I forgot to document) so any per phase-per-guest is out of scope, right?

Reply:
> > Yes, it does. In `__init__.py`, it was put into a "raw step data" for, which is a dictionary one might get from an fmf node, which was then passed to `delegate()` - a plugin is constructed, raw step data normalized, "spec" turns into `Dependency*`. You skip this part, construct the plugin instance on your own, and then inject it into its step data container. The raw step data you have do not get normalized.
> 
> "You skip this part," I'm not, to my knowledge, see
> 
> ```
>     future_recommends: PreparePlugin[Any] = cast(
>         PreparePlugin[Any], PreparePlugin.delegate(
>             prepare_step, raw_data=data_recommend))
>     prepare_step._phases.append(future_recommends)
> ```

I mean here, on line 288: `self.future_requires.data.package = ...`. It's not a key in the raw step data dictionary anymore, IIUIC, it's a `PrepareInstallData.package` attribute, and line 299 stores raw dependency specification in an attribute that's supposed to be `list[tmt.base.DependencySimple]` only. This move is incorrect.

Reply:
thanks for answers, the `uniq` fixed by 9996c7d2. Rest in the next release

Reply:
Thanks, now looks good.